### DISCUSSION TITLE
next start: Port is opened even if the server cannot be started

### DISCUSSION THREAD

**Discussion by timomeh** [vercel/next.js#44819] [Ideas]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.13.1
      npm: 8.1.2
      Yarn: 1.22.15
      pnpm: 6.11.0
    Relevant packages:
      next: 12.1.1-canary.6
      react: 17.0.2
      react-dom: 17.0.2

### How are you deploying your application? (if relevant)

next start

### Describe the Bug

When starting a Next.js server using `next start`, it immediately opens the port. Even if the server crashes right after opening the port, for example due to an error inside `next.config.js`. It always first opens the port, does some initialization (which takes a few seconds based on the app size), then crashes.

This behavior causes some PaaS and Infrastructure-Setups to assume the server is ready and healthy, because they only check if the port was opened. This can cause unhealthy and crashing servers being deployed.

The above scenario with an error inside `next.config.js` happened to me inside a dockerized environment. Inside the `next.config.js`, a file was being required which was available inside the build layer, but not in the final image. This caused the image to build fine, but the server crashed a few seconds after starting the image. Because the next server opened the port immediately and then crashed, the PaaS (GCP CloudRun) did not catch the unhealthy container and it got deployed to production. The PaaS then continued to restart this container again and again, because the opened port indicated that the server started in a healthy state, which it didn't.
If the port would not have been opened, the unhealthy image would not have gotten deployed.

### Expected Behavior

The port should only be opened when the server started in a healthy and ready state.

### To Reproduce

1. Create a new app: `npx create-next-app@latest`
2. Add `require('./bad.js')` to the top of next.config.js
3. Create an empty file `./bad.js`
4. Run `npm run build`  -> build successful
5. Delete `./bad.js`
6. Run `npm start`

Output:

```
ready - started server on 0.0.0.0:3000, url: http://localhost:3000
error - Failed to load next.config.js, see more info here https://nextjs.org/docs/messages/next-config-error
Error: Cannot find module './bad.js'
```

I confirmed that TCP Port 3000 was open for a short amount of time.

**Comment by timomeh** [comment]
This behavior is clearly visible here:

https://github.com/vercel/next.js/blob/0e5e8881c08d2f13f2604327f3940776cc5c6a42/packages/next/cli/next-start.ts#L61-L74
https://github.com/vercel/next.js/blob/0e5e8881c08d2f13f2604327f3940776cc5c6a42/packages/next/server/lib/start-server.ts#L37-L56

`server.listen` should ideally be called after `await app.prepare()`, like the example in the Custom Server Docs: https://nextjs.org/docs/advanced-features/custom-server

If moving `server.listen` after `app.prepare()` wouldn't cause any issues, I'd be happy to submit a PR.

**Comment by jankaifer** [comment]
This thread is quite old, sorry for that.
If you send us a PR we can check if it breaks something.

Maybe a better way to check if the server is healthy is to have a dummy endpoint.
Like `/ping`?

Since it's not really a bug I'll convert this into a discussion.
